### DISCUSSION TITLE
Page router hook does not get the correct query object

### DISCUSSION THREAD

**Discussion by JW9506** [vercel/next.js#12586] [Help] [answered]
## Problem
Page router hooks object does not receive a populated query object.

When I use the `<Link href="..." as="...">` trick to pass parameters to a page, along with Koa listening on the exact route, such as the following, an empty query object is received in the router hooks
![image](https://user-images.githubusercontent.com/34848995/81251588-9ea93100-8ff1-11ea-9d27-079da9810de8.png)

![image](https://user-images.githubusercontent.com/34848995/81251370-1f1b6200-8ff1-11ea-89b5-8fd9ca9523b8.png)

![image](https://user-images.githubusercontent.com/34848995/81251377-23477f80-8ff1-11ea-91d0-1741c5d424a0.png)

## To Reproduce
https://github.com/JW9506/nextjs-bug-reproduce

Issue: the Router hook received in `ZPage` page contains an empty query object
Expected: Whenever ZPage is refreshed, should receive correct query object

The following is observed
![image](https://user-images.githubusercontent.com/34848995/81250734-ae277a80-8fef-11ea-9ffb-b78bfcc66e24.png)

## System information

- OS: [Windows]
- Browser (if applies) [chrome]
- Version of Next.js: [9.3.6]
- Version of Node.js: [v12.16.1]

**Comment by JW9506** [comment]
Can someone help? Or am I doing something wrong in this implementation of url masking?

**Comment by ijjk** [comment, marked as answer]
Hi, the page you are trying to pass the custom `query` values to is [automatically statically optimized](https://nextjs.org/docs/advanced-features/automatic-static-optimization) at build time since you aren't defining a data method such as `getServerSideProps`. 

Since this page is statically optimized to an HTML file at build time instead of re-rendering for each request, the custom `query` values can't be passed using a custom server like this unless you leverage the built-in [dynamic-routing](https://nextjs.org/docs/routing/dynamic-routes) support in Next.js so that the route params can be parsed by Next.js or opt-out of the automatic static optimization by adding `getServerSideProps` to signify that the page requires re-rendering on each request

**Reply by JW9506** [reply]
Thank you, the key thing was I did not specifically add `getInitialProps` method
![image](https://user-images.githubusercontent.com/34848995/81454720-16469f80-915b-11ea-9861-dc07b741fd93.png)
Now, it works

**Comment by jsonjunkie** [comment]
hey, check your repo, i used next router to get access to params instead of link

**Reply by JW9506** [reply]
I checked your pr, you used Router instead of Link but it did not make a difference

**Reply by jsonjunkie** [reply]
hmmm, thats weird. i was able to console.log the param as well as see it on the page after pressing the button on my machine. well glad you got it working anyways!

**Reply by JW9506** [reply]
Did you also try to refresh the new page and see the param rendered on the browser?
